### PR TITLE
Caching for ImageLoader Implemented, LIFO queue logic for thread executer implemented

### PR DESCRIPTION
**Pull request by kulebin**

### CODE REVIEW

File: app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java [hunk 1]
```diff
@@ -207,4 +233,14 @@ private void setPlaceholder(final ImageView pView, final int pResId) {
             pView.setImageResource(android.R.color.transparent);
         }
     }
+
+    private void addBitmapToCache(final String pUrl, final Bitmap pBitmap) {
+        if (getBitmapFromCache(pUrl) == null) {
```
Comment (Mikhail-Ivanou) [inline app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java]: If you reach this method - you will have no image in cache, so it's better just to put twice image to cache in rare edge cases, then to check if image is already in cache every time.

Optimization tip: do logic that is faster for biggest part of user first

File: app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java [hunk 2]
```diff
@@ -30,10 +35,20 @@
...
     private static final int MAX_THREAD_NUMBER = 3;
+    private static final int MAX_CACHE_SIZE = 32 * 1000 * 1000;
+    private static final int CACHE_SIZE = Math.min((int) (Runtime.getRuntime().maxMemory() / 6), MAX_CACHE_SIZE);
 
-    private final ExecutorService mExecutorService = Executors.newFixedThreadPool(MAX_THREAD_NUMBER);
+    private final Executor mExecutor = new ThreadPoolExecutor(MAX_THREAD_NUMBER, MAX_THREAD_NUMBER, 1000, TimeUnit.MILLISECONDS, new LifoBlockingDeque<Runnable>());
     private final Handler mHandler = new Handler(Looper.getMainLooper());
+    private final Lock mLock = new ReentrantLock();
     private ISizeParams mSizeParamsCallback;
+    private final LruCache<String, Bitmap> mLruCache = new LruCache<String, Bitmap>(CACHE_SIZE) {
```
Comment (Mikhail-Ivanou) [inline app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java]: can we have cache as a configurable module?

Comment (kulebin) [reply app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java]: Done

File: app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java (hunk 1 above)
Comment (kulebin) [reply app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java]: Done

**Pull request merged**